examples/step-74/step-laplace-FEIV.cc

```diff
@@ -0,0 +1,1159 @@
+    void   setup_system();
+    void   assemble_system();
+    void   solve();
+    void   refine_grid();
+    void   output_results(const unsigned int cycle) const;
+    double compute_penalty(const double h1, const double h2);
+    void   compute_errors();
+    void   compute_error_estimate();
+    double compute_energy_norm();
+    // The following functions will be moved to FEIV
+    void get_function_values(const FEInterfaceValues<dim> &fe_iv,
+                             const Vector<double> &        solution,
+                             std::vector<double>           face_values[2]);
+    void get_function_jump(const FEInterfaceValues<dim> &fe_iv,
+                           const Vector<double> &        solution,
+                           std::vector<double> &         jump);
+    void get_function_average(const FEInterfaceValues<dim> &fe_iv,
+                              const Vector<double> &        solution,
+                              std::vector<double> &         average);
+    void get_function_gradients(const FEInterfaceValues<dim> &fe_iv,
+                                const Vector<double> &        solution,
+                                std::vector<Tensor<1, dim>> face_gradients[2]);
+    void get_function_gradient_jump(const FEInterfaceValues<dim> &fe_iv,
+                                    const Vector<double> &        solution,
+                                    std::vector<Tensor<1, dim>> &gradient_jump);
+    void get_function_gradient_average(
+      const FEInterfaceValues<dim> &fe_iv,
+      const Vector<double> &        solution,
+      std::vector<Tensor<1, dim>> & gradient_average);
+    const Test_Case      test_case;
+    Triangulation<dim>   triangulation;
+    const MappingQ1<dim> mapping;
+
+    AffineConstraints<double> constraints;
+    using ScratchData = MeshWorker::ScratchData<dim>;
+
+    FE_DGQ<dim>     fe;
+    DoFHandler<dim> dof_handler;
+
+    // The next four members represent the linear system to be
+    // solved. <code>system_matrix</code> and <code>right_hand_side</code> are
+    // generated by <code>assemble_system()</code>, the <code>solution</code>
+    // is computed in <code>solve()</code>. The <code>sparsity_pattern</code>
+    // is used to determine the location of nonzero elements in
+    // <code>system_matrix</code>.
+    SparsityPattern      sparsity_pattern;
+    SparseMatrix<double> system_matrix;
+    Vector<double>       solution;
+    Vector<double>       system_rhs;
+
+
+    Vector<double>   estimated_error_square_per_cell;
+    Vector<double>   energy_norm_square_per_cell;
+    ConvergenceTable convergence_table;
+    const double     viscosity = 1.0;
+  };
+
+
+  // We start with the constructor. The 1 in the constructor call of
+  // <code>fe</code> is the polynomial degree.
+  template <int dim>
+  SIPGLaplace<dim>::SIPGLaplace(const Test_Case &test_case)
+    : test_case(test_case)
+    , mapping()
+    , fe(3)
+    , dof_handler(triangulation)
+  {}
```

Comment:
Your comment mention first order but your constructor uses third order interpolation. Be careful :)!. It might be a good idea to pass the order as a parameter to the constructor?